Fancyauth.APIUtil/FancyPluginBase.cs

```diff
@@ -0,0 +1,349 @@
+            public Func<T, Task> Listener { get; set; }
+
+            public GlobalListener(Func<T, object> a, Func<T, Task> b)
+            {
+                Mapper = a;
+                Listener = b;
+            }
+        }
+        private class ActiveListener<T> : GlobalListener<T>
+        {
+            public object Obj { get; set; }
+
+            public ActiveListener(GlobalListener<T> gl) : base(gl.Mapper, gl.Listener)
+            {
+                Obj = null;
+            }
+        }
+
+        private readonly ConcurrentBag<GlobalListener<IUser>> GlobalUserListeners = new ConcurrentBag<GlobalListener<IUser>>();
+        private readonly ConcurrentDictionary<int, IList<GlobalListener<IUser>>> SingleUserListeners = new ConcurrentDictionary<int, IList<GlobalListener<IUser>>>();
+        private readonly ConcurrentDictionary<int, IList<ActiveListener<IUser>>> UserListeners = new ConcurrentDictionary<int, IList<ActiveListener<IUser>>>();
+
+        private readonly ConcurrentBag<GlobalListener<IChannel>> GlobalChannelListeners = new ConcurrentBag<GlobalListener<IChannel>>();
+        private readonly ConcurrentDictionary<int, IList<GlobalListener<IChannel>>> SingleChannelListeners = new ConcurrentDictionary<int, IList<GlobalListener<IChannel>>>();
+        private readonly ConcurrentDictionary<int, IList<ActiveListener<IChannel>>> ChannelListeners = new ConcurrentDictionary<int, IList<ActiveListener<IChannel>>>();
+
+        public async sealed override Task Init()
+        {
+            bool success = true;
+            foreach (var user in await Server.GetOnlineUsers()) {
+                success &= UserListeners.TryAdd(user.SessionId, new List<ActiveListener<IUser>>());
+            }
+            foreach (var channel in await Server.GetChannels()) {
+                success &= ChannelListeners.TryAdd(channel.ChannelId, new List<ActiveListener<IChannel>>());
+            }
+
+            if (!success) {
+                throw new Exception("Init of FancyPluginBase failed due to non-empty listener-lists or duplicte IDs. " +
+                    "This should actually not happen - we are so screwed!!!");
+            }
+        }
+
+        private void AddGlobalListener<T>(ConcurrentBag<GlobalListener<T>> globalListeners,
+            ConcurrentDictionary<int, IList<ActiveListener<T>>> listeners,
+            Func<T, object> mapper, Func<T, Task> listener) where T : IReadModifyWriteObject
+        {
+            var gl = new GlobalListener<T>(mapper, listener);
+            globalListeners.Add(gl);
+            foreach (var key in listeners.Keys) {
+                listeners.AddOrUpdate(key,
+                    i =>
+                    {
+                        var l = new List<ActiveListener<T>>();
+                        l.Add(new ActiveListener<T>(gl));
+                        return l;
+                    },
+                    (i, l) =>
+                    {
+                        l.Add(new ActiveListener<T>(gl));
```

Comment:
`List<T>` isn't threadsafe, so you can either use the immutable list (I'm doing that *somewhere* in this codebase) or `lock` the list.